build_tools/cmake/iree_tablegen_doc.cmake

```diff
@@ -0,0 +1,75 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(CMakeParseArguments)
+
+# iree_tablegen_doc()
+#
+# Runs iree-tablegen to produce documentation.
+function(iree_tablegen_doc)
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;TBLGEN"
+    "TD_FILE;OUTS"
+    ${ARGN}
+  )
+
+  if(IREE_BUILD_DOCS)
+    # Prefix the library with the package name, so we get: iree_package_name
+    iree_package_name(_PACKAGE_NAME)
+    set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+    if(${_RULE_TBLGEN} MATCHES "IREE")
+      set(_TBLGEN "IREE")
+    else()
+      set(_TBLGEN "MLIR")
+    endif()
+
+
+    set(_INCLUDE_DIRS ${IREE_COMMON_INCLUDE_DIRS})
+    list(APPEND _INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR})
+    list(TRANSFORM _INCLUDE_DIRS PREPEND "-I")
+
+    set(_INPUTS ${_RULE_TD_FILE})
+    set(LLVM_TARGET_DEFINITIONS ${_INPUTS})
+
+    while(_RULE_OUTS)
```

Comment:
I think the `add_custom_command`/`add_custom target` should not be in the while loop! Especially for the `add_custom_target` I am not sure if this is the intended way. One can generate all docs with `tablegen` within the while loop, add those to a list and copy all files in connection with a single target.

Reply:
You are right. Moved it out.